Финальный проект спринта №7

KanID92 opened the pull request:
1) Реализовать ТаскМенеджер FileBackedTaskManager с сохранением в файл, в т.ч.:
- реализовать метод save();
- реализовать enum TaskType;
- реализовать методы String toString(Task task), Task fromString(String value), static String historyToString(HistoryManager manager), static List<Integer> historyFromString(String value); 2) Реализовать собственное непроверяемое исключение ManagerSaveException; 3) Дополнить тесты на реализацию менеджера FileBackedTaskManager;

Корректировки по финальному проекту спринта №6:
- InMemoryHistoryManager. Исключить лишнюю проверку containsValue в методе remove(int id);
- InMemoryHistoryManager. Добавить аннотацию @Override перед методом getHistory().

KanID92 committed:
Финальный проект спринта №7:
1) Реализовать ТаскМенеджер FileBackedTaskManager с сохранением в файл, в т.ч.:
- реализовать метод save();
- реализовать enum TaskType;
- реализовать методы String toString(Task task), Task fromString(String value), static String historyToString(HistoryManager manager), static List<Integer> historyFromString(String value);
2) Реализовать собственное непроверяемое исключение ManagerSaveException;
3) Дополнить тесты на реализацию менеджера FileBackedTaskManager;

Корректировки по финальному проекту спринта №6:
- InMemoryHistoryManager. Исключить лишнюю проверку containsValue в методе remove(int id);
- InMemoryHistoryManager. Добавить аннотацию @Override перед методом getHistory().

kesch9 commented on java-kanban.iml:
```diff
@@ -3,9 +3,27 @@
     </content>
     <orderEntry type="inheritedJdk" />
     <orderEntry type="sourceFolder" forTests="false" />
+    <orderEntry type="module-library" scope="TEST">
```
Нужно добавить папку .idea файл *.iml в gitignore. Эти файлы генерирует idea и их не надо каждый раз коммитить и отдавать на ревью
Вот как вариант можно написать
### IntelliJ IDEA ###
.idea
*.iws
*.iml
*.ipr
или можете посмотреть в интернете еще примеры gitignore

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,332 @@
+
+    private final File autoSaveFile;
+
+    private int counter = 0;
```
Обязательно
Создавать новый счетчик counter не нужно, он должен наследоваться от родителя InMemoryTaskManager, сделайте его protected

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,332 @@
+        this.autoSaveFile = autoSaveFile;
+    }
+
+    private int getId() {
```
Обязательно
Этот метод также не нужен, вы можете его отнаследовать от родителя сделав protected

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,332 @@
+    }
+
+    @Override
+    public ArrayList<SubTask> getSubtaskList() { // геттер списка подзадач
```
Обязательно
Переопределять нужно было методы которые меняют внутреннее состояние таск менеджера. getTaskList, getEpicList, getSubtaskList ничего не меняют, их переопределять не нужно, также вы забыли проставить Override над методом getTaskList

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,332 @@
+    }
+
+    @Override
+    public ArrayList<SubTask> getEpicSubtasks(int epicId) {
```
Обязательно
Этот метод также не нужно было переопределять

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,332 @@
+    }
+
+    @Override
+    public List<Task> getHistory() {
```
Тоже метод не переопределяем

kesch9 replied:
Не подправлено замечание

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,332 @@
+    }
+
+    @Override
+    <T> void deleteTasksFromHistory(HashMap<Integer, T> map) {
```
Обязательно
Этот метод не нужно переопределять, он вспомогательный у вас и вызывается при удалении задач, а эти методы вы переопределили

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,332 @@
+                + task.getDescription() + "," + task.getEpicId();
+    }
+
+    static FileBackedTaskManager loadFromFile(File autoSaveFile) { // восстановление данных менеджера
```
Обязательно
Метод нужно сделать public

kesch9 replied:
Не подправлено замечание

kesch9 commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,332 @@
+                                fileBTManager.tasks.put(task.getId(), task);
+                                break;
+                            case EPIC:
+                                Epic epic = new Epic(task.getName(), task.getDescription());
```
Обязательно
Создавать явно Epic не нужно было можно было скастить
 fileBTManager.epics.put(task.getId(), (Epic) task);

kesch9 commented on test/service/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,102 @@
+        FileBackedTaskManager taskManager4;
+        File file;
+        file = autosaveTestFile;
+        taskManager4 = FileBackedTaskManager.loadFromFile(file);
```
Можно лучше
Не очень эффективно написано и много лишних строчек
FileBackedTaskManager taskManager4 = FileBackedTaskManager.loadFromFile(autosaveTestFile);

kesch9 commented on test/service/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,102 @@
+        File emptyFile;
+        try {
+            emptyFile = File.createTempFile("TestFile2", null);
+            taskManager2 = new FileBackedTaskManager(Managers.getDefaultHistory(), emptyFile);
```
Обязательно
taskManager2 не нужен при проверки чтения из пустого файла, вам нужен только emptyFile

kesch9 reviewed (commented):
Хорошая работа

KanID92 committed:
Финальный проект спринта №7 (корректировка по ревью от 12.03):
- добавить исключения в gitignore;
- убрать лишние методы, не требующие наследования;
- изменить модификатор доступа счетчика;
- оптимизировать код тестов FileBackedTaskManagerTest

kesch9 reviewed (approved):
Approve

kesch9 reviewed (approved):
Approve
